core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -2199,82 +2080,79 @@ object SnappySession extends Logging {
-        // duplicate the cachedDF to avoid overwriting if two similar queries are executed
-        // but collect() invoked later
-        // TODO: SW: handle this case: simple duplicate will not work since two different
-        // executions of same CDF may require two different set of ParamLiterals
-        // so store the actual ParamLiterals from session in duplicated one separately
-        // and change just before execution (internal references array, Filters have reference
-        // to the original ParamLiteral/LiteralValue only and those cannot be duplicated)
-        // cachedDF = cachedDF.duplicate()
-      }
-      handleCachedDataFrame(cachedDF, key, lp, currentWrappedConstants, session, sqlText)
-    } catch {
-      case e: UncheckedExecutionException => e.getCause match {
-        case ee: EntryExistsException =>
-          handleCachedDataFrame(ee.getOldValue.asInstanceOf[CachedDataFrame],
-            key, lp, currentWrappedConstants, session, sqlText)
-        case t => throw t
+      } finally {
+        session.currentKey = null
       }
-      case ee: EntryExistsException =>
-        handleCachedDataFrame(ee.getOldValue.asInstanceOf[CachedDataFrame],
-          key, lp, currentWrappedConstants, session, sqlText)
+    } else {
+      logDebug(s"Using cached plan for: $sqlText (existing: ${cachedDF.queryString})")
+      cachedDF = cachedDF.duplicate()
     }
+    handleCachedDataFrame(cachedDF, key, lp, session, sqlText, paramLiterals)
   }
 
   private def handleCachedDataFrame(cachedDF: CachedDataFrame, key: CachedKey,
-      plan: LogicalPlan, currentWrappedConstants: ArrayBuffer[ParamLiteral],
-      session: SnappySession, sqlText: String): CachedDataFrame = {
-    cachedDF.waitForLastShuffleCleanup()
-    cachedDF.reset()
+      plan: LogicalPlan, session: SnappySession, sqlText: String,
+      currentParamLiterals: Array[ParamLiteral]): CachedDataFrame = {
     cachedDF.queryString = sqlText
-    if (key.valid) {
-      // logDebug(s"calling reprepare broadcast with new constants ${currentWrappedConstants}")
-      // cachedDF.reprepareBroadcast(lp, currentWrappedConstants)
-      logDebug(s"calling replace constants with new constants $currentWrappedConstants" +
-          s" in Literal values = ${cachedDF.allLiterals.toSet}")
-      CachedPlanHelperExec.replaceConstants(cachedDF.allLiterals, plan, currentWrappedConstants)
-    }
-    // set the query hints as would be set at the end of un-cached sql()
-    session.synchronized {
-      session.queryHints.clear()
-      session.queryHints.putAll(cachedDF.queryHints)
+    // replace the tokenized constants with the new ones
+    if (cachedDF.isCached && (cachedDF.paramLiterals eq null)) {
+      cachedDF.paramLiterals = currentParamLiterals
     }
+    cachedDF.currentLiterals = currentParamLiterals
     cachedDF
   }
 
+  /**
+   * Replace any ParamLiterals in a string with current values.
+   */
+  private[sql] def replaceParamLiterals(text: String,
```

Comment:
Will a '#' in the valueString itself affect this logic?

Reply:
A '#' in valueString cannot change the format of ParamLiteral.toString which is like: ParamLiteral(1#4,AB#C). So the '#' after the position integer will not be altered. The length after '#' is written to determine the end of string within instead of searching for ')' for that reason.